src/parser/command.rs

```diff
@@ -213,6 +217,52 @@ pub fn parse_defframe<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instructio
     ))
 }
 
+/// Parse the contents of a `DEFGATE` instruction.
+pub fn parse_defgate<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instruction> {
+    let (input, name) = token!(Identifier(v))(input)?;
+    let (input, parameters) = opt(delimited(
+        token!(LParenthesis),
+        separated_list1(token!(Comma), token!(Variable(v))),
+        token!(RParenthesis),
+    ))(input)?;
+    let (input, gate_type) = opt(preceded(
+        token!(As),
+        alt((
+            map(token!(Matrix), |()| GateType::Matrix),
+            map(token!(Permutation), |()| GateType::Permutation),
+        )),
+    ))(input)?;
+    let (input, _) = token!(Colon)(input)?;
+
+    let gate_type = gate_type.unwrap_or(GateType::Matrix);
+
+    let mut matrix = None;
+    let mut permutation = None;
+    let input = match gate_type {
+        GateType::Matrix => {
+            let (input, m) = parse_matrix(input)?;
+            matrix = Some(m);
+            input
+        }
+        GateType::Permutation => {
+            let (input, p) = parse_permutation(input)?;
+            permutation = Some(p);
+            input
+        }
+    };
```

Comment:
Is there a more idiomatic way to handle this case where I need to initialize two mutually exclusive fields based on a switch variable?

Reply:
```rust
let (input, matrix, permutation) = match gate_type {
    GateType::Matrix => {
        let (input, m) = parse_matrix(input)?;
        (input, Some(m), None)
    }
    GateType::Permutation {
        let (input, p) = parse_permutation(input)?;
        (input, None, Some(p))
    }
};

Reply:
🔥 
It was staring me right in the face...